topics/reference_guide/custom_language_support/language_server_protocol.md

```diff
@@ -0,0 +1,175 @@
+
+## Basic Implementation
+
+A minimal LSP integration must implement `LspServerSupportProvider` along with a service descriptor and [register](plugin_extensions.md#declaring-extensions) it as a `com.intellij.platform.lsp.serverSupportProvider` [Extension Point (EP)](plugin_extension_points.md):
+
+```kotlin
+import com.intellij.platform.lsp.api.LspServerSupportProvider
+import com.intellij.platform.lsp.api.ProjectWideLspServerDescriptor
+
+class FooLspServerSupportProvider : LspServerSupportProvider {
+  override fun fileOpened(project: Project, file: VirtualFile, serverStarter: LspServerStarter) {
+    if (file.extension == "foo") {
+      serverStarter.ensureServerStarted(FooLspServerDescriptor(project))
+    }
+  }
+}
+
+private class FooLspServerDescriptor(project: Project) : ProjectWideLspServerDescriptor(project, "Foo") {
+  override fun isSupportedFile(file: VirtualFile) = file.extension == "foo"
+  override fun createCommandLine() = GeneralCommandLine("foo", "--stdio")
+}
+```
+
+After providing the implementation of the LSP Server support provider extension point, register it in the <path>plugin.xml</path> file as follows:
+
+```xml
+<idea-plugin>
+   <!-- ... -->
+   <depends>com.intellij.modules.platform</depends>
+   <depends>com.intellij.modules.ultimate</depends>
+   <extensions defaultExtensionNs="com.intellij">
+       <platform.lsp.serverSupportProvider implementation="FooLspServerSupportProvider"/>
+   </extensions>
+</idea-plugin>
+```
+
+As a reference, check out the [Prisma ORM](https://plugins.jetbrains.com/plugin/20686-prisma-orm) open-source plugin implementation: https://github.com/JetBrains/intellij-plugins/tree/master/prisma/src/org/intellij/prisma/ide/lsp
+
+## Language Server Integration
+
+Language Server is a separate process that analyzes source code and provides language-specific features to development tools.
+When creating a plugin that utilizes LSP within the IDE, there are two possibilities for providing a Language Server to end-users:
+
+- Bundle a [Language Server implementation](https://microsoft.github.io/language-server-protocol/implementors/servers/) binary as a resource delivered with a plugin.
+- Provide a possibility for users to define the location of the Language Server binary in their environment.
+
+The Prisma ORM plugin presents the first approach, which distributes the `prisma-language-server.js` script and uses a local Node.js interpreter to run it.
+
+For more complex cases, the plugin may request to provide a detailed configuration with a dedicated [Settings](settings_guide.md) implementation.
+
+To implement a minimal LSP plugin, perform the following steps:
+- Within the `LspServerSupportProvider.fileOpened()` method, spin up the relevant LSP server descriptor, which can decide if the given file is supported by using the `LspServerDescriptor.isSupportedFile()` check method.
+- Tell how to start the server by implementing `LspServerDescriptor.createCommandLine()`.
```

Comment:
I think this would be more natural in the `Basic Implementation` section, before the code sample that illustrates exactly these 2 points.